"""PSI-MI 25 Molecular interaction types (https://www.ebi.ac.uk/ols/ontologies/mi)."""

types = {0: 'molecular interaction', 1: 'interaction detection method', 2: 'participant identification method',
         3: 'feature detection method', 4: 'affinity chromatography technology', 5: 'alanine scanning',
         6: 'anti bait coimmunoprecipitation', 7: 'anti tag coimmunoprecipitation', 8: 'array technology',
         9: 'bacterial display', 10: 'beta galactosidase complementation', 11: 'beta lactamase complementation',
         12: 'bioluminescence resonance energy transfer', 13: 'biophysical', 14: 'adenylate cyclase complementation',
         16: 'circular dichroism', 17: 'classical fluorescence spectroscopy', 18: 'two hybrid',
         19: 'coimmunoprecipitation',
         20: 'transmission electron microscopy', 21: 'colocalization by fluorescent probes cloning',
         22: 'colocalization by immunostaining', 23: 'colocalization/visualisation technologies',
         24: 'confirmational text mining', 25: 'copurification', 26: 'correlated mutations', 27: 'cosedimentation',
         28: 'cosedimentation in solution', 29: 'cosedimentation through density gradient', 30: 'cross-linking study',
         31: 'protein cross-linking with a bifunctional reagent',
         32: 'de novo protein sequencing by mass spectrometry',
         33: 'deletion analysis', 34: 'display technology', 35: 'docking', 36: 'domain fusion',
         37: 'domain profile pairs',
         38: 'dynamic light scattering', 39: 'edman degradation', 40: 'electron microscopy',
         41: 'electron nuclear double resonance', 42: 'electron paramagnetic resonance', 43: 'electron resonance',
         45: 'experimental interaction detection', 46: 'experimental knowledge based', 47: 'far western blotting',
         48: 'filamentous phage display', 49: 'filter binding', 50: 'flag tag coimmunoprecipitation',
         51: 'fluorescence technology', 52: 'fluorescence correlation spectroscopy',
         53: 'fluorescence polarization spectroscopy', 54: 'fluorescence-activated cell sorting',
         55: 'fluorescent resonance energy transfer', 56: 'full identification by DNA sequencing',
         57: 'gene neighbourhood',
         58: 'genome based prediction', 59: 'gst pull down', 60: 'ha tag coimmunoprecipitation', 61: 'his pull down',
         62: 'his tag coimmunoprecipitation', 63: 'interaction prediction', 64: 'interologs mapping',
         65: 'isothermal titration calorimetry', 66: 'lambda phage display', 67: 'light scattering',
         68: 'mass detection of residue modification', 69: 'mass spectrometry studies of complexes',
         70: 'mobility shift',
         71: 'molecular sieving', 72: 'monoclonal antibody western blot', 73: 'mrna display', 74: 'mutation analysis',
         75: 'myc tag coimmunoprecipitation', 76: 'neural network on interface properties',
         77: 'nuclear magnetic resonance',
         78: 'nucleotide sequence identification', 79: 'other biochemical technologies',
         80: 'partial DNA sequence identification by hybridization', 81: 'peptide array',
         82: 'peptide massfingerprinting',
         83: 'peptide synthesis', 84: 'phage display', 85: 'phylogenetic profile',
         86: 'polyclonal antibody western blot',
         87: 'predictive text mining', 88: 'primer specific pcr', 89: 'protein array',
         90: 'protein complementation assay',
         91: 'chromatography technology', 92: 'protein in situ array', 93: 'protein sequence identification',
         94: 'protein staining', 95: 'proteinchip(r) on a surface-enhanced laser desorption/ionization',
         96: 'pull down',
         97: 'reverse ras recruitment system', 98: 'ribosome display', 99: 'scintillation proximity assay',
         100: 'sequence based phylogenetic profile', 101: 'sequence based prediction',
         102: 'sequence tag identification',
         103: 'southern blot', 104: 'static light scattering', 105: 'structure based prediction',
         106: 'surface patches',
         107: 'surface plasmon resonance', 108: 't7 phage display', 109: 'tap tag coimmunoprecipitation',
         110: 'text mining',
         111: 'dihydrofolate reductase reconstruction', 112: 'ubiquitin reconstruction', 113: 'western blot',
         114: 'x-ray crystallography', 115: 'yeast display', 116: 'feature type', 117: 'binding-associated region',
         118: 'mutation', 119: 'mutation decreasing interaction', 120: 'post translation modification',
         121: 'acetylated residue', 122: 'n-acetyl-alanine', 123: 'n2-acetyl-arginine', 124: 'n-acetyl-asparagine',
         125: 'n-acetyl-aspartic acid', 126: 'n-acetyl-cysteine', 127: 'n-acetyl-glutamine',
         128: 'n-acetyl-glutamic acid',
         129: 'n-acetylglycine', 130: 'n-acetyl-histidine', 131: 'n-acetyl-isoleucine', 132: 'n-acetyl-leucine',
         133: 'n2-acetyl-lysine', 134: 'n6-acetyl-lysine', 135: 'n-acetyl-methionine', 136: 'n-acetyl-phenylalanine',
         137: 'n-acetyl-proline', 138: 'n-acetyl-serine', 139: 'n-acetyl-threonine', 140: 'n-acetyl-tryptophan',
         141: 'n-acetyl-tyrosine', 142: 'n-acetyl-valine', 143: 'amidated residue', 145: 'arginine amide',
         146: 'formylated residue',
         147: 'n-formyl-methionine', 148: 'hydroxylated residue', 150: 'lipid modification',
         151: 's-farnesyl-cysteine', 152: 's-geranylgeranyl-cysteine', 153: 'n-palmitoyl-cysteine',
         154: 's-palmitoyl-cysteine',
         155: 'n-myristoyl-glycine', 156: 'n6-myristoyl-lysine', 157: 'methylated residue', 158: 'n-methyl-alanine',
         159: 'n,n,n-trimethyl-alanine', 160: 'omega-n,omega-n-dimethyl-arginine',
         161: 'beta-methylthioaspartic acid',
         162: 'n5-methyl-glutamine', 163: 'glutamic acid 5-methyl ester', 165: 'n6-methyl-lysine',
         166: 'n6,n6-dimethyl-lysine',
         167: 'n6,n6,n6-trimethyl-lysine', 168: 'n-methyl-methionine', 169: 'n-methyl-phenylalanine',
         170: 'phosphorylated residue', 171: 'omega-n-phospho-arginine', 172: 'aspartic 4-phosphoric anhydride',
         173: 's-phospho-cysteine', 176: 'o-phospho-serine', 177: 'o-phospho-threonine', 178: "o4'-phospho-tyrosine",
         179: 'other modification', 180: 'selenocysteine', 181: 'selenomethionine', 182: '3-oxoalanine',
         184: 'glutamyl 5-glycerylphosphorylethanolamine', 186: 'n6-biotinyl-lysine',
         187: 'n6-(4-amino-2-hydroxybutyl)-lysine',
         188: 'n6-retinal-lysine', 189: 'ubiquitinated lysine', 190: 'interaction type', 191: 'aggregation',
         192: 'acetylation reaction', 193: 'amidation reaction', 194: 'cleavage reaction', 195: 'covalent binding',
         196: 'covalent interaction', 197: 'deacetylation reaction', 198: 'defarnesylation reaction',
         199: 'deformylation reaction', 200: 'degeranylation reaction', 201: 'demyristoylation reaction',
         202: 'depalmitoylation reaction', 203: 'dephosphorylation reaction', 204: 'deubiquitination reaction',
         205: 'disaggregation', 206: 'farnesylation reaction', 207: 'formylation reaction',
         208: 'genetic interaction',
         209: 'geranylgeranylation reaction', 210: 'hydroxylation reaction', 211: 'lipid addition',
         212: 'lipoprotein cleavage reaction', 213: 'methylation reaction', 214: 'myristoylation reaction',
         215: 'non covalent interaction', 216: 'palmitoylation reaction', 217: 'phosphorylation reaction',
         218: 'physical interaction', 219: 'synthetic lethal', 220: 'ubiquitination reaction',
         221: 'expression level',
         222: 'physiological level', 223: 'under expressed level', 225: 'chromatin immunoprecipitation array',
         226: 'ion exchange chromatography', 227: 'reverse phase chromatography',
         228: 'cytoplasmic complementation assay',
         230: 'membrane bound complementation assay', 231: 'mammalian protein protein interaction trap',
         232: 'transcriptional complementation assay', 233: 'protein dna complex', 234: '131i radiolabel',
         235: '14c radiolabel', 236: '32p radiolabel', 237: '33p radiolabel', 238: '3h radiolabel',
         239: 'biotin tag',
         240: 'fusion protein', 241: 'horseradish peroxidase tag', 242: 'gene ontology definition reference',
         243: 'isoform parent sequence reference', 244: 'reactome complex', 245: 'reactome protein', 246: 'cabri',
         247: 'newt',
         248: 'resid', 249: 'huge', 250: 'gene', 251: 'gene product', 252: 'biological feature',
         253: 'isotope label',
         254: 'genetic interference', 255: 'post transcriptional interference', 256: 'rna interference',
         257: 'antisense rna',
         258: 'inhibitor antibodies', 259: 'perturbagens peptides', 260: 'inhibitor small molecules',
         261: 'suppression',
         262: 'suppression mutation', 263: 'suppression knockout', 264: 'suppression partial alteration',
         265: 'suppression expression alteration', 266: 'suppression overexpression', 267: 'suppression scalable',
         268: 'suppression underexpression', 269: 'synthetic phenotype', 270: 'conditional synthetic lethal',
         271: 'conditional synthetic lethal temperature-sensitivity', 273: 'synthetic growth effect',
         274: 'synthetic growth defect', 275: 'synthetic growth increase', 276: 'blue native page',
         300: 'alias type',
         301: 'gene name', 302: 'gene name synonym', 303: 'gene ontology synonym', 304: 'isoform synonym',
         305: 'ordered locus name', 306: 'open reading frame name', 307: 'delivery method', 308: 'electroporation',
         309: 'genomic tagging', 310: 'infection', 311: 'microinjection', 312: 'nucleic acid transfection',
         313: 'interactor type', 314: 'complex', 315: 'protein complex', 316: 'ribonucleoprotein complex',
         317: 'interaction',
         318: 'nucleic acid', 319: 'deoxyribonucleic acid', 320: 'ribonucleic acid', 321: 'catalytic rna',
         322: 'guide rna',
         323: 'heterogeneous nuclear rna', 324: 'messenger rna', 325: 'transfer rna', 326: 'protein',
         327: 'peptide',
         328: 'small molecule', 329: 'unknown participant', 330: 'molecular source', 331: 'engineered',
         332: 'naturally occurring', 333: 'feature range status', 334: 'c-terminal position',
         335: 'certain sequence position',
         336: 'greater-than', 337: 'less-than', 338: 'range', 339: 'undetermined sequence position',
         340: 'n-terminal position',
         341: 'ragged n-terminus', 342: 'sample process', 343: 'cdna library', 344: 'cell lysate',
         345: 'author assigned name',
         346: 'experimental preparation', 348: 'fixed cell', 349: 'living cell', 350: 'purified',
         351: 'homogeneous',
         352: 'partially purified', 353: 'cross-reference type', 354: 'cellular component',
         355: 'molecular function',
         356: 'identical object in an external resource', 357: 'method reference', 358: 'primary-reference',
         359: 'biological process', 360: 'secondary accession number', 361: 'additional information',
         362: 'inference',
         363: 'inferred by author', 364: 'inferred by curator', 365: 'enzyme tag', 366: 'alkaline phosphatase tag',
         367: 'green fluorescent protein tag', 368: 'yellow fluorescent protein tag',
         369: 'lex-a dimerization assay',
         370: 'tox-r dimerization assay', 371: '35s radiolabel', 372: 'subcellular preparation', 373: 'dye label',
         374: 'cyanine label', 375: 'cy3 label', 376: 'cy5 label', 377: 'fluorescein isothiocyanate label',
         378: 'rare isotope label', 379: '13c label', 380: '15n label', 381: '2h label',
         382: 'mutation increasing interaction',
         383: 'biopolymer', 384: 'alexa label', 385: 'alexa 350 label', 386: 'alexa 430 label',
         387: 'alexa 488 label',
         388: 'alexa 532 label', 389: 'alexa 546 label', 390: 'alexa 568 label', 391: 'alexa 594 label',
         396: 'predetermined participant', 397: 'two hybrid array', 398: 'two hybrid pooling approach',
         399: 'two hybrid fragment pooling approach', 400: 'affinity technology', 401: 'biochemical',
         402: 'chromatin immunoprecipitation assay', 403: 'colocalization',
         404: 'comigration in non denaturing gel electrophoresis', 405: 'competition binding',
         406: 'deacetylase assay',
         407: 'direct interaction', 408: 'disulfide bond', 409: 'dna footprinting', 410: 'electron tomography',
         411: 'enzyme linked immunosorbent assay', 412: 'electrophoretic mobility supershift assay',
         413: 'electrophoretic mobility shift assay', 414: 'enzymatic reaction', 415: 'enzymatic study',
         416: 'fluorescence microscopy', 417: 'footprinting', 418: 'genetic', 419: 'gtpase assay',
         420: 'kinase homogeneous time resolved fluorescence', 421: 'identification by antibody',
         422: 'immunostaining',
         423: 'in-gel kinase assay', 424: 'protein kinase assay', 425: 'kinase scintillation proximity assay',
         426: 'light microscopy', 427: 'Identification by mass spectrometry', 428: 'imaging technique',
         429: 'necessary binding region', 430: 'nucleic acid uv cross-linking assay', 431: 'obsolete',
         432: 'one hybrid',
         433: 'partial identification of protein sequence', 434: 'phosphatase assay', 435: 'protease assay',
         436: 'protein footprinting', 437: 'protein three hybrid', 438: 'rna three hybrid',
         439: 'random spore analysis',
         440: 'saturation binding', 441: 'synthetic genetic analysis', 442: 'sufficient binding region',
         443: 'ubiquitin binding', 444: 'database citation', 445: 'literature database', 446: 'pubmed',
         447: 'feature database',
         448: 'gene ontology', 449: 'interpro', 450: 'cdd', 451: 'pfam', 452: 'pirsf', 453: 'prints', 454: 'prodom',
         455: 'prosite', 456: 'scop superfamily', 457: 'smart', 458: 'tigrfams', 459: 'mmdb', 460: 'rcsb pdb',
         461: 'interaction database', 462: 'bind', 463: 'biogrid', 464: 'cygd', 465: 'dip', 466: 'ecocyc',
         467: 'reactome',
         468: 'hprd', 469: 'intact', 470: 'kegg', 471: 'mint', 472: 'pdbe', 473: 'participant database', 474: 'chebi',
         475: 'ddbj/embl/genbank', 476: 'ensembl', 477: 'entrez gene/locuslink', 478: 'flybase', 479: 'mgd/mgi',
         480: 'omim',
         481: 'refseq', 482: 'rfam', 483: 'rgd', 484: 'sgd', 485: 'uniparc', 486: 'uniprot knowledge base',
         487: 'wormbase',
         488: 'psi-mi', 489: 'source database', 490: 'experiment condition', 491: 'in silico', 492: 'in vitro',
         493: 'in vivo',
         494: 'in situ', 495: 'experimental role', 496: 'bait', 497: 'neutral component', 498: 'prey',
         499: 'unspecified role',
         500: 'biological role', 501: 'enzyme', 502: 'enzyme target', 503: 'self', 505: 'experimental feature',
         506: 'over expressed level', 507: 'tag', 508: 'deacetylase radiometric assay',
         509: 'phosphatase homogeneous time resolved fluorescence', 510: 'homogeneous time resolved fluorescence',
         511: 'protease homogeneous time resolved fluorescence', 512: 'zymography', 513: 'collagen film assay',
         514: 'in gel phosphatase assay', 515: 'methyltransferase assay', 516: 'methyltransferase radiometric assay',
         517: 'radiolabel', 518: 'flag tag', 519: 'glutathione s tranferase tag', 520: 'ha tag', 521: 'his tag',
         522: 'myc tag',
         523: 't7 tag', 524: 'calmodulin binding peptide plus protein a tag', 525: 'v5 tag', 526: 'n-acetyl-lysine',
         527: 'adp ribosylated residue', 528: 'omega-n-(adp-ribosyl)-arginine', 529: 's-(adp-ribosyl)-cysteine',
         530: 'glutamyl-5-poly(adp-ribose)', 531: 'o-(adp-ribosyl)-serine', 532: 'n4-(adp-ribosyl)-asparagine',
         533: 'glycosylated residue', 534: 'glycosyl-cysteine', 535: 'glycosyl-serine', 536: 'glycosyl-threonine',
         537: 'omega-n-glycosyl-arginine', 538: 'n4-glycosyl-asparagine', 539: 'gpi anchor residue',
         540: 'gpi-anchor amidated alanine', 541: 'gpi-anchor amidated asparagine',
         542: 'gpi-anchor amidated aspartate',
         543: 'gpi-anchor amidated cysteine', 544: 'gpi-anchor amidated glycine', 545: 'gpi-anchor amidated serine',
         546: 'gpi-anchor amidated threonine', 547: 's-prenyl-cysteine', 548: 'methylated-lysine',
         549: 'alkylated cysteine',
         550: 'gamma-carboxyglutamic acid', 551: 'nitro-tyrosine', 552: 's-nitrosyl-cysteine',
         553: "o4'-sulfo-tyrosine",
         554: 'sumoylated lysine', 555: 'phospho-histidine', 556: 'transglutamination reaction',
         557: 'adp ribosylation reaction', 558: 'deglycosylation reaction', 559: 'glycosylation reaction',
         560: 'myristoylated residue', 561: 'palmitoylated residue', 562: 'methylated alanine',
         563: 'methylated arginine',
         564: 'omega-n-methyl-arginine', 565: 'neddylated lysine', 566: 'sumoylation reaction',
         567: 'neddylation reaction',
         568: 'desumoylation reaction', 569: 'deneddylation reaction', 570: 'protein cleavage', 571: 'mrna cleavage',
         572: 'dna cleavage', 573: 'mutation disrupting interaction', 574: 'digital object identifier',
         575: 'alliance for cellular signaling', 576: 'structural proximity',
         577: 'feature prediction from structure',
         578: 'maltose binding protein tag', 579: 'electron donor', 580: 'electron acceptor', 581: 'suppressor gene',
         582: 'suppressed gene', 583: 'fluorescence donor', 584: 'fluorescence acceptor', 585: 'intenz',
         586: 'inhibitor',
         587: 'inhibited', 588: 'three hybrid', 589: 'in vitro translated protein', 590: 'attribute name',
         591: 'experiment description', 592: 'ipfam', 593: 'translocation', 594: 'translocation start',
         595: 'translocation end', 596: 'experimental form description', 597: 'feature description',
         598: 'feature constraint',
         599: 'figure legend', 600: 'conditional synthetic lethal nutrition-sensitivity', 601: 'sequence ontology',
         602: 'chemical footprinting', 603: 'dimethylsulphate footprinting',
         604: 'potassium permanganate footprinting',
         605: 'enzymatic footprinting', 606: 'DNase I footprinting', 607: 'small nuclear rna', 608: 'ribosomal rna',
         609: 'small nucleolar rna', 610: 'small interfering rna', 611: 'signal recognition particle rna',
         612: 'comment',
         613: 'function', 614: 'url', 615: 'search-url', 616: 'example', 617: 'disease', 618: 'caution',
         619: 'pathway',
         620: 'search-url-ascii', 621: 'author-confidence', 622: 'confidence-mapping', 623: 'inhibition',
         624: 'stimulant',
         625: 'agonist', 626: 'antagonist', 627: 'experiment modification', 628: 'validation regular expression',
         629: 'complex-properties', 630: '3d-structure', 631: '3d-r-factors', 632: '3d-resolution',
         633: 'data-processing',
         634: 'contact-email', 635: 'contact-comment', 636: 'author-list', 637: 'isoform-comment',
         638: 'prerequisite-ptm',
         639: 'resulting-ptm', 640: 'parameter type', 641: 'ic50', 642: 'ec50', 643: 'ki', 644: 'km', 645: 'kcat',
         646: 'Kd',
         647: 'parameter unit', 648: 'molar', 649: 'second', 650: 'millimolar', 651: 'micromolar', 652: 'nanomolar',
         653: 'picomolar', 654: 'fentomolar', 655: 'lambda repressor two hybrid', 656: 'identified peptide',
         657: 'systematic evolution of ligands by exponential enrichment',
         658: 'multidimensional protein identification technology', 659: 'experimental feature detection',
         660: 'feature prediction', 661: 'experimental participant identification', 662: 'imex-primary',
         663: 'confocal microscopy', 664: 'interaction attribute name', 665: 'experiment attibute name',
         666: 'participant attribute name', 667: 'controlled vocabulary attribute name',
         668: 'feature attribute name',
         669: 'organism attribute name', 670: 'imex', 671: 'antibodies', 672: 'library-used', 673: 'complex synonym',
         674: 'peptide parent sequence reference', 675: 'international protein index',
         676: 'tandem affinity purification',
         677: 'tandem tag', 678: 'antibody array', 679: 'poly adenine', 680: 'single stranded deoxyribonucleic acid',
         681: 'double stranded deoxyribonucleic acid', 682: 'cofactor', 683: 'sequence database', 684: 'ancillary',
         685: 'source reference', 686: 'unspecified method', 687: 'fluorescent protein tag',
         688: 'dna binding domain tag',
         689: 'activation domain tag', 690: 'gal4 activation domain', 691: 'vp16 activation domain',
         692: 'b42 activation domain', 693: 'gal4 dna binding domain', 694: 'lexa dna binding domain',
         695: 'sandwich immunoassay', 696: 'polymerase assay', 697: 'dna directed dna polymerase assay',
         698: 'dna directed rna polymerase assay', 699: 'rna directed dna polymerase assay',
         700: 'rna directed rna polymerase assay', 701: 'dna strand elongation', 702: 'panther', 703: 'gene3d',
         704: 'nucleic acid delivery', 705: 'anti tag western blot', 706: 'nucleic acid transformation',
         707: 'anti tag immunostaining', 708: 'monoclonal antibody immunostaining',
         709: 'polyclonal antibody immunostaining',
         710: 'nucleic acid transformation by treatment with divalent cation', 711: 'nucleic acid electroporation',
         712: 'nucleic acid microinjection', 713: 'nucleic acid passive uptake', 714: 'nucleic acid transduction',
         715: 'nucleic acid conjugation', 716: 'passive uptake', 717: 'nucleic acid transfection with liposome',
         718: 'nucleic acid transfection by treatment', 719: 'calcium phosphate nucleic acid transfection',
         720: 'nucleic acid delivery by infection', 721: 'protein delivery', 722: 'protein electroporation',
         723: 'protein microinjection', 724: 'protein delivery by cationic lipid treatment',
         725: 'protein delivery by infection', 726: 'reverse two hybrid', 727: 'lexa b52 complementation',
         728: 'gal4 vp16 complementation', 729: 'luminescence based mammalian interactome mapping', 730: 'pubchem',
         731: '3d repertoire', 732: 'red fluorescent protein tag', 733: 'cyan fluorescent protein tag',
         734: 'enhanced green fluorescent protein tag', 735: 'transactivating tag', 736: 'protein passive uptake',
         737: 'peptide sequence database', 738: 'pride', 739: 'penetrating tag', 740: 'cell penetrating peptide tag',
         741: 'peptide atlas', 742: 'gpm', 787: 'genetic experimental form', 788: 'knock out', 789: 'knock down',
         790: 'hypermorph', 791: 'hypomorph', 792: 'antimorph', 793: 'amorph', 794: 'synthetic', 795: 'asynthetic',
         796: 'suppression', 797: 'epistasis', 798: 'conditional genetic interaction defined by inequality',
         799: 'additive genetic interaction defined by inequality',
         800: 'single nonmonotonic genetic interaction defined by inequality',
         801: 'double nonmonotonic genetic interaction defined by inequality', 802: 'enhancement interaction',
         803: 'expression level alteration', 804: 'mutated gene', 805: 'wwpdb', 806: 'pdbj',
         807: 'comigration in gel electrophoresis', 808: 'comigration in sds page',
         809: 'bimolecular fluorescence complementation', 810: 'substitution analysis', 811: 'insertion analysis',
         812: 'calmodulin binding protein tag', 813: 'proximity ligation assay',
         814: 'protease accessibility laddering',
         815: 'confirmation by molecular weight', 816: 'molecular weight estimation by staining',
         817: 'molecular weight estimation by silver staining',
         818: 'molecular weight estimation by coomasie staining',
         819: 'molecular weight estimation by bromide staining', 820: 'molecular weight estimation by sybr staining',
         821: 'molecular weight estimation by autoradiography',
         822: 'molecular weight estimation by hoechst staining',
         823: 'predetermined feature', 824: 'x-ray powder diffraction', 825: 'x-ray fiber diffraction',
         826: 'x ray scattering',
         827: 'x-ray tomography', 828: 'polyprotein fragment', 829: 'multiple parent reference', 830: 'tissue list',
         831: 'cell ontology', 832: 'half cystine', 833: 'autoradiography', 834: 'ka', 835: 'koff',
         836: 'temperature of interaction', 837: 'pH of interaction', 838: 'kelvin', 839: 'per mole per second',
         840: 'stimulator', 841: 'phosphotransferase assay', 842: 'phosphate donor', 843: 'phosphate acceptor',
         844: 'phosphotransfer reaction', 845: 'spin label', 846: 'r1 spin label', 847: 'dansyl label',
         848: '125i radiolabel',
         849: 'ncbi taxonomy', 850: 'encode', 851: 'protein genbank identifier', 852: 'nucleotide genbank identifier',
         853: 'dna overhang', 854: '3 prime overhang', 855: '5 prime overhang', 856: 'fluorophore',
         857: 'fluorescent dye label', 858: 'immunodepleted coimmunoprecipitation', 859: 'intermolecular force',
         860: 'genbank indentifier', 861: 'protein a tag', 862: 'zz tag', 863: 'thiol reactive lanthanide label',
         864: 'brenda',
         865: 'fluorescence acceptor donor pair', 866: 'tag visualisation', 867: 'tag visualisation by fluorescence',
         868: 'author identifier', 869: 'originally assigned identifier', 870: 'demethylase assay',
         871: 'demethylation reaction', 872: 'atomic force microscopy', 873: 'curation request', 875: 'dataset',
         878: 'author submitted', 879: 'nucleoside triphosphatase assay', 880: 'atpase assay',
         881: 'nucleoside triphosphatase reaction', 882: 'atpase reaction', 883: 'gtpase reaction', 884: 'vsv tag',
         885: 'journal', 886: 'publication year', 887: 'histone acetylase assay',
         888: 'small angle neutron scattering',
         889: 'acetylase assay', 890: 'qdot', 891: 'neutron fiber diffraction', 892: 'solid phase assay',
         893: 'neutron diffraction', 894: 'electron diffraction', 895: 'protein kinase A complementation',
         896: 'renilla luciferase protein tag', 897: 'protein modification ontology', 898: 'putative self',
         899: 'p3 filamentous phage display', 900: 'p8 filamentous phage display', 901: 'isotope label footprinting',
         902: 'rna cleavage', 903: 'mpidb', 904: 'polysaccharide',
         905: 'amplified luminescent proximity homogeneous assay',
         906: 'au1 tag', 907: 'conformational status', 908: 'denatured', 909: 'native', 910: 'nucleic acid cleavage',
         911: 'cross linker', 912: 'spdp cross linker', 913: 'lc-spdp cross linker', 914: 'association',
         915: 'physical association', 916: 'lexa vp16 complementation', 917: 'matrixdb', 918: 'donor', 919: 'acceptor',
         920: 'ribonuclease assay', 921: 'surface plasmon resonance array', 922: 'imex evidence', 923: 'irefindex',
         924: 'camjedb', 925: 'observed-ptm', 926: 'fiash label', 927: 'iaedans label', 928: 'filter trap assay',
         929: 'northern blot', 930: 'epistatis', 931: 'genetic interaction defined by inequality',
         932: 'noninteractive',
         933: 'negative genetic interaction', 934: 'neutral genetic interaction',
         935: 'positive genetic interaction',
         936: 'emdb', 937: 'glu tag', 938: 'rheology measurement', 939: 'fluorescein label',
         940: 'fluorescein-5-maleimide label', 941: 'competitor', 942: 'uniprot taxonomy',
         943: 'detection by mass spectrometry', 944: 'mass spectrometry study of hydrogen/deuterium exchange',
         945: 'oxidoreductase activity electron transfer reaction', 946: 'miniaturized immunoprecipitation',
         947: 'bead aggregation assay', 948: 'kinetic conditions', 949: 'gdp/gtp exchange assay',
         950: 'trapping mutant',
         951: 'chain parent sequence reference', 952: 'imex secondary', 953: 'polymerization',
         954: 'curation quality',
         955: 'curation depth', 956: 'curation coverage', 957: 'full coverage', 958: 'partial coverage',
         959: 'imex curation',
         960: 'mimix curation', 961: 'rapid curation', 962: 'strep ii tag',
         963: 'interactome parallel affinity capture',
         964: 'infrared spectroscopy', 965: '2d-infrared spectrometry', 966: 'ultraviolet-visible spectroscopy',
         967: 'chembl compound', 968: 'biosensor', 969: 'bio-layer interferometry', 970: 'inchi key',
         971: 'phosphopantetheinylation', 972: 'phosphopantetheinylase assay', 973: 'imex source', 974: 'innatedb',
         975: 'fc-igg tag', 976: 'total internal reflection fluorescence spectroscopy', 977: 'no-imex-export',
         978: 'author-name', 979: 'oxidoreductase assay', 980: 'tag visualisation by enzyme assay',
         981: 'tag visualisation by peroxidase activity', 982: 'electrophoretic mobility-based method', 983: 'gemma',
         984: 'deamination assay', 985: 'deamination reaction', 986: 'nucleic acid strand elongation reaction',
         987: 'rna strand elongation', 988: 'strep tag', 989: 'amidase assay', 990: 'cleavage assay',
         991: 'lipoprotein cleavage assay', 992: 'defarnesylase assay', 993: 'degeranylase assay',
         994: 'demyristoylase assay',
         995: 'depalmitoylase assay', 996: 'deformylase assay', 997: 'ubiquitinase assay', 998: 'deubiquitinase assay',
         999: 'formylase assay', 1000: 'hydroxylase assay', 1001: 'lipidase assay', 1002: 'myristoylase assay',
         1003: 'geranylgeranylase assay', 1004: 'palmitoylase assay', 1005: 'adp ribosylase assay',
         1006: 'deglycosylase assay',
         1007: 'glycosylase assay', 1008: 'sumoylase assay', 1009: 'desumoylase assay', 1010: 'neddylase assay',
         1011: 'deneddylase assay', 1012: 'sbp', 1013: 'ensemblgenomes', 1014: 'string', 1015: 'dictybase',
         1016: 'fluorescence recovery after photobleaching', 1017: 'rna immunoprecipitation', 1018: 'deltamass',
         1019: 'protein phosphatase assay', 1020: 'hilyte fluor 488', 1021: 'qx 520',
         1022: 'field flow fractionation',
         1023: 'luminogreen', 1024: 'scanning electron microscopy', 1025: 'unimod', 1026: 'diphtamidase assay',
         1027: 'diphtamidation reaction', 1028: 'modified chromatin immunoprecipitation',
         1029: 'proteomics of isolated chromatin segments', 1030: 'excimer fluorescence',
         1031: 'protein folding/unfolding',
         1032: 'atto 488', 1033: 'atto 550', 1034: 'nuclease assay', 1035: 'deoxyribonuclease assay',
         1036: 'nucleotide exchange assay', 1037: 'Split renilla luciferase complementation',
         1038: 'silicon nanowire field-effect transistor', 1039: 'c-terminal range', 1040: 'n-terminal range',
         1041: 'synonym',
         1042: 'pubmed central', 1043: 'flannotator', 1044: 'rice genome annotation project',
         1045: 'curation content',
         1046: 'interacting molecules', 1047: 'protein-protein', 1048: 'smallmolecule-protein',
         1049: 'nucleicacid-protein',
         1050: 'interaction representation', 1051: 'evidence', 1052: 'clustered', 1053: 'data source',
         1054: 'experimentally-observed', 1055: 'internally-curated', 1056: 'text-mining', 1057: 'predicted',
         1058: 'imported',
         1059: 'complex expansion', 1060: 'spoke expansion', 1061: 'matrix expansion', 1062: 'bipartite expansion',
         1063: 'consensuspathdb', 1064: 'interaction confidence', 1065: 'replication-based confidence',
         1066: 'structure-based confidence', 1067: 'function-based confidence', 1068: 'location-based confidence',
         1069: 'network-based confidence', 1070: 'standard-based confidence', 1071: 'literature-based confidence',
         1072: 'method-based confidence', 1073: 'statistical-based confidence', 1074: 'rgs-his tag',
         1075: 'beilstein',
         1076: 'einecs', 1077: 'merck index', 1078: 'plantgdb', 1079: 'ratmap', 1080: 'tair', 1081: 'tigr/jcvi',
         1082: 'zfin',
         1083: 'cog', 1084: 'photon donor', 1085: 'photon acceptor', 1086: 'equilibrium dialysis',
         1087: 'monoclonal antibody blockade', 1088: 'phenotype-based detection assay',
         1089: 'nuclear translocation assay',
         1090: 'bimane label', 1091: 'publication title', 1092: 'atto label', 1093: 'bibliographic attribute name',
         1094: 'genome databases', 1095: 'hgnc', 1096: 'protein sequence databases', 1097: 'uniprot',
         1098: 'uniprot/swiss-prot', 1099: 'uniprot/trembl', 1100: 'bioactive entity', 1101: 'standard inchi key',
         1102: 'mapped-identity', 1103: 'solution state nmr', 1104: 'solid state nmr', 1105: 'biocyc',
         1106: 'pathways database', 1107: 'pid', 1108: 'biocarta', 1109: 'gene database',
         1110: 'predicted interaction',
         1111: 'two hybrid bait or prey pooling approach', 1112: 'two hybrid prey pooling approach',
         1113: 'two hybrid bait and prey pooling approach', 1114: 'virhostnet', 1115: 'spike', 1116: 'genemania',
         1117: 'topfind', 1118: 'enhanced yellow fluorescent protein tag', 1119: 'nYFP', 1120: 'cYFP', 1121: 'ceYFP',
         1122: 'neYFP', 1123: 'bindingdb', 1124: 'pathwaycommons', 1125: 'direct binding region',
         1126: 'self interaction',
         1127: 'putative self interaction', 1128: 'mutation disrupting interaction strength',
         1129: 'mutation disrupting interaction rate', 1130: 'mutation decreasing interaction rate',
         1131: 'mutation increasing interaction rate', 1132: 'mutation increasing interaction strength',
         1133: 'mutation decreasing interaction strength', 1134: 'mcherry fluorescent protein tag',
         1135: 'venus fluorescent protein tag', 1136: 'kusabira-green protein tag', 1137: 'carboxylation assay',
         1138: 'decarboxylation assay', 1139: 'carboxylation reaction', 1140: 'decarboxylation reaction',
         1141: 's tag',
         1142: 'aminoacylation assay', 1143: 'aminoacylation reaction', 1144: 'protein a tag visualisation',
         1145: 'phospholipase assay', 1146: 'phospholipase reaction', 1147: 'ampylation assay',
         1148: 'ampylation reaction',
         1149: 'cooperative interaction', 1150: 'affected interaction', 1151: 'participant-ref',
         1152: 'cooperative effect value', 1153: 'cooperative effect outcome', 1154: 'positive cooperative effect',
         1155: 'negative cooperative effect', 1156: 'cooperative mechanism', 1157: 'allostery', 1158: 'pre-assembly',
         1159: 'allosteric molecule', 1160: 'allosteric effector', 1161: 'allosteric response',
         1162: 'allosteric k-type response', 1163: 'allosteric v-type response', 1164: 'allosteric mechanism',
         1165: 'allosteric change in structure', 1166: 'allosteric change in dynamics', 1167: 'allostery type',
         1168: 'heterotropic allostery', 1169: 'homotropic allostery', 1170: 'pre-assembly response',
         1171: 'composite binding site formation', 1172: 'altered physicochemical compatibility',
         1173: 'binding site hiding',
         1174: 'configurational pre-organization', 1175: 'allosteric post-translational modification',
         1176: 'sequence based prediction of gene regulatory region binding sites',
         1177: 'phylogenetic profiling of predicted gene regulatory region binding sites',
         1178: 'sequence based prediction of binding of transcription factor to transcribed gene regulatory elements',
         1179: 'partial nucleotide sequence identification', 1180: 'partial DNA sequence identification',
         1181: 'paired end tags sequence identification', 1182: 'full identification by RNA sequencing',
         1183: 'nuclease footprinting', 1184: 'dna adenine methyltransferase identification',
         1185: 'tag visualisation by dna adenine methyltransferase', 1186: 'dna methyltransferase tag', 1187: 'damip',
         1188: 'tag visualisation by mutated dna adenine methyltransferase', 1189: 'methylation interference assay',
         1190: 'hydroxy radical footprinting', 1191: 'ultraviolet (uv) footprinting',
         1192: 'antisense oligonucleotides',
         1193: 'partial RNA sequence identification', 1194: 'reverse transcription pcr', 1195: 'quantitative pcr',
         1196: 'quantitative reverse transcription pcr', 1197: 'radioimmunoassay', 1198: 'immunohistochemistry',
         1199: 'anti-tag immunohistochemistry', 1200: 'immunocytochemistry', 1201: 'anti-tag immunocytochemistry',
         1202: 'one-strep-tag', 1203: 'split luciferase complementation',
         1204: 'split firefly luciferase complementation',
         1205: 'luciferase tag', 1206: 'renilla-n', 1207: 'renilla-c', 1208: 'firefly luciferase protein tag',
         1209: 'firefly-c', 1210: 'firefly-n', 1211: 'liposome binding assay', 1212: 'checksum', 1213: 'n-venus',
         1214: 'c-venus', 1215: 'bifc tag', 1216: 'cGFP', 1217: 'nGFP', 1218: 'chromosome conformation capture assay',
         1219: 'enzyme-mediated activation of radical sources', 1220: 'tag visualisation by luciferase assay',
         1221: 'author-based confidence', 1222: 'mbinfo', 1223: 'ptm decreasing an interaction',
         1224: 'ptm increasing an interaction', 1225: 'ptm disrupting an interaction', 1226: 'ampylation assay',
         1227: 'lap tag', 1228: 'pyo tag', 1229: 'uridylation assay', 1230: 'uridylation reaction',
         1231: 'aminomethylcoumarin label', 1232: 'aggregation assay', 1233: 'resulting-cleavage', 1234: 'silac',
         1235: 'thermal shift binding', 1236: 'proline isomerase assay', 1237: 'proline isomerization  reaction',
         1238: 'mass spectrometry studies of subunit exchange', 1239: 'amino-acid variant',
         1240: 'disease causing amino-acid variant', 1241: 'variant', 1242: 'fc-igg1', 1243: 'fc-igg2', 1244: 'mkate2',
         1245: 'mkate', 1246: 'ion mobility mass spectrometry of complexes', 1247: 'microscale thermophoresis',
         1248: 'bodipy label', 1249: 'isomerase assay', 1250: 'isomerase reaction',
         1251: 'methylmalonyl-CoA isomerase reaction', 1252: 'methylmalonyl-CoA isomerase asf say', 1253: 'atto 532',
         1254: 'atto 647', 1255: 'stilbene label', 1256: 'luminscent dye label', 1257: 'rhodamine label',
         1258: 'tetramethyl rhodamine label', 1259: 'acrylodan label', 1260: 'pyrene label',
         1261: 'oregon green label',
         1262: 'iid', 1263: 'molecular connections', 1264: 'ntnu', 1265: 'ubiquitin reconstruction tag', 1266: 'cub',
         1267: 'nub', 1268: 'nubg', 1269: 'duplicated protein', 1270: 'xpress tag', 1271: 'enhancement',
         1272: 'positive epistasis', 1273: 'maximal epistasis', 1274: 'minimal epistasis', 1275: 'neutral epistasis',
         1276: 'opposing epistasis', 1277: 'qualitative epistasis', 1278: 'mutual enhancement',
         1279: 'unilateral enhancement',
         1280: 'mutual suppression', 1281: 'mutual suppression (complete)', 1282: 'mutual suppression (partial)',
         1283: 'suppression-enhancement', 1284: 'quantitative epistasis', 1285: 'opposing epistasis',
         1286: 'over-suppression',
         1287: 'mutual over-suppression', 1288: 'over-suppression-enhancement', 1289: 'phenotype bias',
         1290: 'suppression (complete)', 1291: 'suppression (partial)', 1292: 'unilateral suppression',
         1293: 'unilateral suppression (complete)', 1294: 'unilateral suppression (partial)',
         1295: 'unilateral over-suppression', 1296: 'amino acid analysis', 1297: 'phosphoamino acid analysis',
         1298: 'complex type', 1299: 'complex composition', 1300: 'obligate complex', 1301: 'non-obligate complex',
         1302: 'stable complex', 1303: 'transient complex', 1304: 'molecule set', 1305: 'candidate set',
         1306: 'open set',
         1307: 'defined set', 1308: 'resulting sequence', 1309: 'de-ADP-ribosylation assay',
         1310: 'de-ADP-ribosylation reaction', 1311: 'differential scanning calorimetry', 1312: 'aut-page',
         1313: 'proximity labelling technology', 1314: 'proximity-dependent biotin identification',
         1315: 'complex recommended name', 1316: 'complex systematic name', 1317: 'eukaryotic linear motif resource',
         1318: 'sulfate donor', 1319: 'sulfate acceptor', 1320: 'membrane yeast two hybrid',
         1321: 'ire1 reconstruction',
         1322: 'atto 465', 1323: 'tag visualisation by alkaline phosphatase activity', 1324: 'conditioned medium',
         1325: 'sulfurtransferase assay', 1326: 'CLONE OF phosphotransfer reaction', 1327: 'sulfurtransfer reaction',
         1328: 'coumarin label', 1329: 'cpm', 1330: 'dnp', 1331: 'evidence ontology', 1332: 'bhf-ucl', 1333: 'rogid',
         1334: 'rigid', 1335: 'hpidb', 1336: 'experiment database', 1337: 'efo', 1338: 'eef tag',
         1339: 'supercharged green fluorescent protein', 1340: 'human orfeome collection', 1341: 'set member',
         1342: 'qcmd',
         1343: 'enzyme regulator', 1344: 'erythrosin iodoacetamide label', 1345: 'rho tag', 1346: 'bmrb',
         1347: 'protein ontology', 1348: 'chembl target', 1349: 'chembl', 1350: 'orphanet', 1351: 'inferred-from',
         1352: 'uracil interference assay', 1353: 'au5 tag', 1354: 'lipase assay', 1355: 'lipid cleavage',
         1356: 'validated two hybrid', 1357: 'RNAcentral', 2002: 'drugbank', 2003: 'commercial name',
         2004: 'drug brand name',
         2005: 'drug mixture brand name', 2006: 'biotech product preparation', 2007: 'iupac name',
         2008: 'chemical formula',
         2009: 'chemical structure', 2010: 'standard inchi', 2011: 'cas registry number', 2012: 'kegg compound',
         2013: 'pubchem', 2015: 'pharmgkb', 2016: 'bind smid', 2017: 'heterogen',
         2020: 'canadian drug identification number',
         2021: 'rxlist link', 2023: 'material safety data sheet', 2024: 'patent number', 2025: 'molecular weight',
         2026: 'melting point', 2027: 'water solubility', 2029: 'logp', 2030: 'isoelectric point',
         2033: 'hydrophobicity',
         2036: 'boiling point', 2039: 'smiles string', 2040: 'drug type', 2041: 'drug category',
         2042: 'disease indication',
         2043: 'pharmacology', 2044: 'mechanism of action', 2045: 'drug absorption', 2046: 'lethal dose 50',
         2047: 'percentage of plasma protein binding', 2048: 'drug biotransformation', 2049: 'elimination half life',
         2050: 'dosage form', 2051: 'patient information', 2053: 'contraindications',
         2054: 'bioactive entity reference',
         2055: 'chemical stability', 2064: 'solubility', 2084: 'organisms affected',
         2086: 'physicochemical attribute name',
         2089: 'bioactive entity attribute name', 2091: 'structure representation attribute name',
         2097: 'anti-convulsant',
         2098: 'anti-bacterial', 2099: 'fda approved drug', 2100: 'experimental drug', 2101: 'biotech drug',
         2102: 'nutraceutical drug', 2105: 'pka', 2106: 'degree of ionisation ph 7.4', 2107: 'logd',
         2108: 'solubility ph 7.4',
         2109: 'solubility in dmso', 2111: 'diffusion coefficient', 2112: 'chemical stability at pH 2',
         2113: 'dissolution profile', 2115: 'pharmacokinetics attribute name', 2116: 'cell permeability',
         2118: 'volume of distribution', 2120: 'tissue distribution', 2121: 'transporter binding', 2122: 'clearance',
         2123: 'renal clearance', 2124: 'total clearance', 2125: 'maximum absorbable dose',
         2126: 'paracellular absorption',
         2127: 'tmax/cmax', 2128: 'ABCB1 transporter substrate', 2129: 'bile transporter substrate',
         2130: 'cyp-450 inhibition',
         2131: 'metabolite identification', 2132: 'gsh adducts',
         2133: 'neutralization by glucuronidation or sulfatation',
         2135: 'toxicity attribute name', 2136: 'herg binding', 2137: 'genotoxicity', 2138: 'mutagenicity',
         2139: 'carcinogenicity', 2140: 'chromosome damage', 2141: 'hepatotoxicity', 2142: 'phospholipidosis',
         2145: 'solubility ph 6.5', 2146: 'solubility ph 2.0', 2147: 'chemical stability at pH 7.4',
         2148: 'investigational drug', 2149: 'withdrawn drug', 2150: 'illicit drug', 2151: 'other drug interaction',
         2152: 'food interaction', 2153: 'pdr health', 2154: 'wikipedia', 2155: 'average molecular weight',
         2156: 'monoisotopic molecular weight', 2157: 'experimental water solubility',
         2158: 'predicted water solubility',
         2160: 'logs', 2161: 'experimental logs', 2162: 'experimental CaCO2 permeability', 2163: 'by homology',
         2164: 'mind',
         2165: 'bar', 2166: 'ai', 2167: 'kinetic exclusion assay', 2168: 'conditional site labelling',
         2169: 'luminiscence technology', 2170: 'bimolecular luminiscence complementation',
         2171: 'complemented donor-acceptor resonance energy transfer', 2172: 'aspgd', 2173: 'cgd', 2174: 'ecoliwiki',
         2175: 'genedb', 2176: 'gramene', 2177: 'pombase', 2178: 'agi_locuscode', 2179: 'subset', 2180: 'agbase',
         2181: 'cacao',
         2182: 'dflat', 2183: 'go_central', 2184: 'mtbbase', 2185: 'parkinsonsuk-ucl', 2186: 'alut', 2187: 'ri',
         2188: 'par-clip', 2189: 'avexis', 2190: 'long non-coding ribonucleic acid', 2191: 'clip', 2192: 'clip-seq',
         2193: 'iclip', 2194: 'crac', 2195: 'clash', 2196: 'quartz crystal microbalance',
         2197: 'probe interaction assay',
         2198: 'labelling assay', 2199: 'specific site-labelling technology', 2200: 'primesdb',
         2201: 'DNA chemical modification', 2202: 'RNA chemical modification', 2203: 'primer extension assay',
         2204: 'micro rna', 2205: 'pir', 2206: 'observed nucleic acid chemical modification',
         2207: 'resulting nucleic acid chemical modification', 2208: 'prerequisite-nucleic acid chemical modification',
         2209: 'nucleic acid chemical modification decreasing an interaction',
         2210: 'nucleic acid chemical modification disrupting an interaction',
         2211: 'nucleic acid chemical modification increasing an interaction', 2212: 'proteomexchange',
         2213: 'super-resolution microscopy', 2214: 'signor', 2215: 'barcode fusion genetics two hybrid',
         2216: 'deampylation assay', 2217: 'luciferase-c', 2218: 'luciferase-n',
         2219: 'gaussia luciferase protein tag',
         2220: 'gaussia-c', 2221: 'gaussia-n', 2222: 'inference by socio-affinity scoring',
         2223: 'inference by quantitative co-purification',
         2224: 'chemical rna modification plus base pairing prediction',
         2225: 'zinc', 2226: 'mutation with no effect', 2227: 'mutation causing an interaction', 2228: 'ceitec',
         2229: 'nucleicacid-gene', 2230: 'nucleicacid-nucleicacid', 2231: 'coexpression',
         2232: 'molecular association',
         2233: 'causal interaction', 2234: 'causal statement', 2235: 'up-regulates', 2236: 'up-regulates activity',
         2237: 'up-regulates quantity', 2238: 'up-regulates quantity by expression',
         2239: 'up-regulates quantity by stabilization', 2240: 'down-regulates', 2241: 'down-regulates activity',
         2242: 'down-regulates quantity', 2243: 'down-regulates quantity by repression',
         2244: 'down-regulates quantity by destabilization', 2245: 'causal regulatory mechanism',
         2246: 'indirect causal regulation', 2247: 'transcriptional regulation', 2248: 'translation regulation',
         2249: 'post transcriptional regulation', 2250: 'direct causal regulation',
         2251: 'transcriptional regulation by direct binding of dbTF to DNA regulatory element',
         2252: 'guanine nucleotide exchange factor reaction', 2253: 'gtpase-activating protein reaction',
         2254: 'chemical activation reaction', 2255: 'chemical inhibition reaction', 2256: 'relocalization',
         2257: 'small molecule catalysis reaction', 2258: 'xenobiotic', 2259: 'causal interactor type',
         2260: 'stimulus',
         2261: 'phenotype', 2262: 'causal regulatory modification', 2263: 's-nitrosylation',
         2264: 'tyrosinated residue',
         2265: 'de-acetylated residue', 2266: 'de-phosphorylated residue', 2267: 'de-sumoylated residue',
         2268: 'de-methylated residue', 2269: 'de-ubiquitinylated residue', 2270: 'signalink', 2271: 'edam',
         2272: 'tyrosinylation', 2273: 'tyrosination', 2274: 'regulator', 2275: 'regulator target',
         2276: 'carbohydrate chemical modification', 2277: 'Cr-two hybrid', 2278: 'polymer chain length',
         2279: 'complex portal', 2280: 'deamidation reaction', 2281: 'deamidation assay', 2282: 'complex-primary',
         2283: 'southwestern blotting', 2284: 'complex component',
         2285: 'miRNA interference luciferase reporter assay',
         2286: 'functional association', 2287: 'identification by structure determination', 2288: 'DAP-seq'}
